### DISCUSSION TITLE
Main advantage of using pandas-ai

### DISCUSSION THREAD

**Discussion by shantanuo** [sinaptik-ai/pandas-ai#589] [Q&A] [answered]
I am getting similar results using this code.
Is there any advantage of using your library?
```
!pip install langchain
import os
os.environ["OPENAI_API_KEY"] = "XXXX"

from langchain.agents import create_pandas_dataframe_agent
from langchain.llms import OpenAI

pd_agent = create_pandas_dataframe_agent(OpenAI(temperature=0), df, verbose=True)

pd_agent.run("Find the total sales for each product line in the year 2003")
```

**Comment by gventuri** [comment, marked as answer]
@shantanuo good question. Generally each library/framework can be better or worse depending on the use case.

In particular, PandasAI:
- doesn't use Chain of Thought, meaning that we only send 1 request to the LLM, instead of multiple, which makes it generally faster and less expensive
- allows also to plot charts
- is tailored for the data analysis, meaning that it's built to deal with structured data

Depending on your use case, this could be a better library or not. My suggestion is to give it a try and check which one is your preferred way to go :)

Let me know if you have any question!

**Comment by astro-siddhesh** [comment]
hello Shantanuo , have you tried to plot graphs using code you provided. I guess it wont work. If you able to code in that way , please share it with me

**Comment by shantanuo** [comment]
Yes, I can see that graph and other advance and sophisticated features are not there in langchain. I appreciate this effort. 
How do I see the raw commands executed on pandas dataframe?

**Comment by gventuri** [comment]
@shantanuo just run `sdf.last_code_executed` where `sdf` is the smart dataframe or smart datalake!

Let me know if that works :)

**Comment by shantanuo** [comment]
That worked. Thanks.
Is there any way to know expected expenses? I am using openAI API

**Comment by daniel-counto** [comment]
@gventuri but seems like the verbose = True is not working? I have started a post in the Bug report.
